File: tools/migrate_play_routes.mjs
import { readFileSync, writeFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const DEFAULT_PROGRESSION = ["WR1", "WR2", "TE1", "RB1", "RB2", "TE2", "WR3"];

function formationRoles(form) {
  return (form.positions || []).map((p) => p.role).filter(Boolean);
}

function pickCarrier(roles) {
  return roles.find((r) => r.toUpperCase().startsWith("RB")) || "";
}

function filteredProgression(roles) {
  const out = [];
  for (const want of DEFAULT_PROGRESSION) {
    if (roles.includes(want)) out.push(want);
  }
  for (const r of roles) {
    const ru = r.toUpperCase();
    if ((ru.startsWith("WR") || ru.startsWith("TE") || ru.startsWith("RB")) && !out.includes(r)) {
      out.push(r);
    }
  }
  return out;
}

function chain(steps) {
  return steps.map((s) => [s[0], s[1]]);
}

function wpsRunCarrier() {
  return chain([[0, -1], [0, -1], [0, -1], [-1, -1], [0, -1], [0, -1]]);
}

function wpsRunLeadBlock() {
  return chain([[0, -1], [1, 0], [0, -1]]);
}

function wpsRunClearOut(role) {
  const lat = role.includes("1") || role.endsWith("3") ? -1 : 1;
  return chain([[0, -1], [lat, -1], [0, -1]]);
}

function wpsPassFlat(role) {
  let n = 1;
  if (role.toUpperCase().startsWith("RB") && role.length > 2) {
    n = parseInt(role.slice(2), 10) || 1;
  }
  const lat = n % 2 === 1 ? -1 : 1;
  return chain([[lat, 0], [lat, -1], [0, -1], [0, -1]]);
}

function wpsPassReceiver(idx, role) {
  if (role.toUpperCase().startsWith("TE")) {
    return chain([[0, -1], [0, -1], [1, -1], [0, -1]]);
  }
  if (idx % 3 === 0) {
    return chain([[0, -1], [0, -2], [0, -1], [1, -1], [0, -1]]);
  }
  if (idx % 3 === 1) {
    return chain([[0, -1], [1, 0], [0, -1], [0, -2]]);
  }
  return chain([[1, -1], [0, -1], [0, -1], [-1, -1]]);
}

function defaultRoleAssignments(ptype, roles) {
  const ra = {};
  const carrier = ptype === "run" ? pickCarrier(roles) : "";
  let rbI = 0;
  let wrI = 0;
  for (const role of roles) {
    const ru = role.toUpperCase();
    if (ru.startsWith("OL") || ru === "QB") {
      ra[role] = { start_action: ptype === "pass" ? "pass_block" : "run_block" };
    } else if (ru.startsWith("RB")) {
      rbI += 1;
      if (ptype === "run") {
        ra[role] = { start_action: role === carrier ? "carry" : "run_block" };
      } else {
        ra[role] = { start_action: "route" };
      }
    } else if (ru.startsWith("WR") || ru.startsWith("TE")) {
      wrI += 1;
      if (ptype === "run") {
        ra[role] = { start_action: wrI % 2 === 0 ? "run_block" : "route" };
      } else {
        ra[role] = { start_action: "route" };
      }
    }
  }
  return ra;
}

function defaultRoutes(ptype, roles) {
  const routes = {};
  const carrier = ptype === "run" ? pickCarrier(roles) : "";
  let rbI = 0;
  let wrI = 0;
  const ra = defaultRoleAssignments(ptype, roles);
  for (const role of roles) {
    const ru = role.toUpperCase();
    let wps = [];
    if (ru.startsWith("RB")) {
      rbI += 1;
      if (ptype === "run" && role === carrier) wps = wpsRunCarrier();
      else if (ptype === "run") wps = wpsRunLeadBlock();
      else if (ptype === "pass") wps = wpsPassFlat(role);
    } else if (ru.startsWith("WR") || ru.startsWith("TE")) {
      wrI += 1;
      if (ptype === "run" && ra[role]?.start_action === "route") {
        wps = wpsRunClearOut(role);
      } else if (ptype === "pass") {
        wps = wpsPassReceiver(wrI, role);
      }
    }
    if (wps.length) routes[role] = { waypoints: wps };
  }
  return routes;
}

function enrichPlay(play, form) {
  const ptype = play.play_type;
  if (ptype !== "run" && ptype !== "pass") return play;
  const roles = formationRoles(form);
  if (!roles.length) return play;
  play.role_assignments = defaultRoleAssignments(ptype, roles);
  play.routes = defaultRoutes(ptype, roles);
  if (ptype === "run") {
    play.ball_carrier_role = pickCarrier(roles);
  } else {
    const order = filteredProgression(roles);
    play.receiver_progression = {};
    if (order[0]) play.receiver_progression.primary = order[0];
    if (order[1]) play.receiver_progression.secondary = order[1];
    if (order[2]) play.receiver_progression.tertiary = order[2];
    play.qb_script = {
      mode: "dropback_progression",
      dropback_ticks: 5,
      progression: order,
    };
  }
  return play;
}

const playsPath = join(ROOT, "data", "plays.json");
const formsPath = join(ROOT, "data", "formations.json");
const plays = JSON.parse(readFileSync(playsPath, "utf8"));
const formsRaw = JSON.parse(readFileSync(formsPath, "utf8"));
const formsList = Array.isArray(formsRaw) ? formsRaw : formsRaw.formations || [];
const forms = Object.fromEntries(formsList.map((f) => [f.id, f]));
let n = 0;
for (const [pid, play] of Object.entries(plays)) {
  if (play.side !== "offense") continue;
  if (play.play_type !== "run" && play.play_type !== "pass") continue;
  const form = forms[play.formation_id];
  if (!form) {
    console.log(`skip ${pid}: unknown formation ${play.formation_id}`);
    continue;
  }
  enrichPlay(play, form);
  n += 1;
}
writeFileSync(playsPath, JSON.stringify(plays, null, "\t") + "\n", "utf8");
console.log(`migrated ${n} offense run/pass plays`);

File: tools/migrate_play_routes.py
#!/usr/bin/env python3
"""One-shot: add routes / role_assignments / progression to offense run & pass plays."""
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_PROGRESSION = ["WR1", "WR2", "TE1", "RB1", "RB2", "TE2", "WR3"]


def formation_roles(form):
    return [p["role"] for p in form.get("positions", []) if "role" in p]


def pick_carrier(roles):
    for r in roles:
        if r.upper().startswith("RB"):
            return r
    return ""


def filtered_progression(roles):
    out = []
    for want in DEFAULT_PROGRESSION:
        if want in roles:
            out.append(want)
    for r in roles:
        ru = r.upper()
        if (ru.startswith("WR") or ru.startswith("TE") or ru.startswith("RB")) and r not in out:
            out.append(r)
    return out


def chain(steps):
    return [[int(s[0]), int(s[1])] for s in steps]


def wps_run_carrier():
    return chain([[0, -1]] * 3 + [[-1, -1], [0, -1], [0, -1]])


def wps_run_lead_block():
    return chain([[0, -1], [1, 0], [0, -1]])


def wps_run_clear_out(role):
    lat = -1 if "1" in role or role.endswith("3") else 1
    return chain([[0, -1], [lat, -1], [0, -1]])


def wps_pass_flat(role):
    n = 1
    if role.upper().startswith("RB") and len(role) > 2:
        try:
            n = int(role[2:])
        except ValueError:
            n = 1
    lat = -1 if n % 2 == 1 else 1
    return chain([[lat, 0], [lat, -1], [0, -1], [0, -1]])


def wps_pass_receiver(idx, role):
    if role.upper().startswith("TE"):
        return chain([[0, -1], [0, -1], [1, -1], [0, -1]])
    if idx % 3 == 0:
        return chain([[0, -1], [0, -2], [0, -1], [1, -1], [0, -1]])
    if idx % 3 == 1:
        return chain([[0, -1], [1, 0], [0, -1], [0, -2]])
    return chain([[1, -1], [0, -1], [0, -1], [-1, -1]])


def default_role_assignments(ptype, roles):
    ra = {}
    carrier = pick_carrier(roles) if ptype == "run" else ""
    rb_i = wr_i = 0
    for role in roles:
        ru = role.upper()
        if ru.startswith("OL") or ru == "QB":
            ra[role] = {"start_action": "pass_block" if ptype == "pass" else "run_block"}
        elif ru.startswith("RB"):
            rb_i += 1
            if ptype == "run":
                ra[role] = {"start_action": "carry" if role == carrier else "run_block"}
            else:
                ra[role] = {"start_action": "route"}
        elif ru.startswith("WR") or ru.startswith("TE"):
            wr_i += 1
            if ptype == "run":
                ra[role] = {"start_action": "run_block" if wr_i % 2 == 0 else "route"}
            else:
                ra[role] = {"start_action": "route"}
    return ra


def default_routes(ptype, roles):
    routes = {}
    carrier = pick_carrier(roles) if ptype == "run" else ""
    rb_i = wr_i = 0
    ra = default_role_assignments(ptype, roles)
    for role in roles:
        ru = role.upper()
        wps = []
        if ru.startswith("RB"):
            rb_i += 1
            if ptype == "run" and role == carrier:
                wps = wps_run_carrier()
            elif ptype == "run":
                wps = wps_run_lead_block()
            elif ptype == "pass":
                wps = wps_pass_flat(role)
        elif ru.startswith("WR") or ru.startswith("TE"):
            wr_i += 1
            if ptype == "run" and ra[role]["start_action"] == "route":
                wps = wps_run_clear_out(role)
            elif ptype == "pass":
                wps = wps_pass_receiver(wr_i, role)
        if wps:
            routes[role] = {"waypoints": wps}
    return routes


def enrich_play(play, form):
    ptype = play.get("play_type", "")
    if ptype not in ("run", "pass"):
        return play
    roles = formation_roles(form)
    if not roles:
        return play
    play["role_assignments"] = default_role_assignments(ptype, roles)
    play["routes"] = default_routes(ptype, roles)
    if ptype == "run":
        play["ball_carrier_role"] = pick_carrier(roles)
    else:
        order = filtered_progression(roles)
        play["receiver_progression"] = {}
        if len(order) > 0:
            play["receiver_progression"]["primary"] = order[0]
        if len(order) > 1:
            play["receiver_progression"]["secondary"] = order[1]
        if len(order) > 2:
            play["receiver_progression"]["tertiary"] = order[2]
        play["qb_script"] = {
            "mode": "dropback_progression",
            "dropback_ticks": 5,
            "progression": order,
        }
    return play


def main():
    plays_path = ROOT / "data" / "plays.json"
    forms_path = ROOT / "data" / "formations.json"
    plays = json.loads(plays_path.read_text(encoding="utf-8"))
    forms = {f["id"]: f for f in json.loads(forms_path.read_text(encoding="utf-8"))["formations"]}
    n = 0
    for pid, play in plays.items():
        if play.get("side") != "offense":
            continue
        if play.get("play_type") not in ("run", "pass"):
            continue
        fid = play.get("formation_id", "")
        form = forms.get(fid)
        if not form:
            print(f"skip {pid}: unknown formation {fid}")
            continue
        enrich_play(play, form)
        n += 1
    plays_path.write_text(json.dumps(plays, indent="\t") + "\n", encoding="utf-8")
    print(f"migrated {n} offense run/pass plays")


if __name__ == "__main__":
    main()
